runtime/src/main/kotlin/kotlin/native/concurrent/SharedReference.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package kotlin.native.concurrent
+
+import kotlin.native.internal.*
+
+@SymbolName("Kotlin_SharedReference_createSharedRef")
+external private fun createSharedRef(value: Any): NativePtr
+
+@SymbolName("Kotlin_SharedReference_derefSharedRef")
+external private fun derefSharedRef(ref: NativePtr): Any
+
+/**
+ * A frozen shared reference to a Kotlin object
+ *
+ * Can be safely passed between workers, but can only be accessed on the worker it was created on,
+ * unless the referred object is frozen too
+ *
+ * Note: Garbage collector currently cannot free any reference cycles with [SharedReference] or [DisposableSharedReference] in them.
+ * Consider using [DisposableSharedReference] to manually resolve cycles with explicit calls to [DisposableSharedReference.dispose]
+ */
+@Frozen
+@NoReorderFields
+@ExportTypeInfo("theSharedReferenceTypeInfo")
+public class SharedReference<out T : Any>(value: T) {
+
+    private val ptr = createSharedRef(value)
+
+    /**
+     * The referenced value.
+     * @throws IncorrectDereferenceException if referred object is not frozen and this is called from a different worker, than [SharedReference] was created on
```

Comment:
If the access fails with exception, it is extremely useful to report some additional information, for example the name of the worker owning this reference and the name of current worker. (Information about the object should still be included).

Supporting this ad hoc for `SharedReference` (instead of `KRefSharedHolder`) is ok for now.
Note that `Worker.name` doesn't work for terminated workers, so the name of the owner should probably be stored to the reference when it is created.

Reply:
(Making exceptions more verbose here can also be postponed to next PR).

Reply:
Implemented in https://github.com/JetBrains/kotlin-native/pull/4115